src/osd/OSD.cc

```diff
@@ -9713,131 +9582,158 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
-      dout(20) << __func__ << slot.to_process.back()
-	       << " queued, waiting_for_pg" << dendl;
+    uint64_t requeue_seq = slot->requeue_seq;
+    ++slot->num_running;
+
+    sdata->sdata_op_ordering_lock.Unlock();
+    osd->service.maybe_inject_dispatch_delay();
+    pg->lock();
+    osd->service.maybe_inject_dispatch_delay();
+    sdata->sdata_op_ordering_lock.Lock();
+
+    auto q = sdata->pg_slots.find(token);
+    if (q == sdata->pg_slots.end()) {
+      // this can happen if we race with pg removal.
+      dout(20) << __func__ << " slot " << token << " no longer there" << dendl;
+      pg->unlock();
       sdata->sdata_op_ordering_lock.Unlock();
       return;
     }
-    pg = slot.pg;
-    dout(20) << __func__ << " " << slot.to_process.back()
-	     << " queued" << dendl;
-    ++slot.num_running;
-  }
-  sdata->sdata_op_ordering_lock.Unlock();
+    slot = q->second.get();
+    --slot->num_running;
 
-  osd->service.maybe_inject_dispatch_delay();
+    if (slot->pg != pg) {
+      // this can happen if we race with pg removal.
+      dout(20) << __func__ << " slot " << token << " no longer attached to "
+	       << pg << dendl;
+      pg->unlock();
+      pg = slot->pg;
+    }
 
-  // [lookup +] lock pg (if we have it)
-  while (true) {
-    if (!pg) {
-      pg = osd->_lookup_lock_pg(token);
-    } else {
-      pg->lock();
-      if (pg->is_deleted()) {
-	dout(20) << __func__ << " got deleted pg " << pg << ", retrying" << dendl;
+    if (slot->to_process.empty()) {
+      // raced with wake_pg_waiters or consume_map
+      dout(20) << __func__ << " " << token
+	       << " nothing queued" << dendl;
+      if (pg) {
 	pg->unlock();
-	pg = nullptr;
-	continue;
       }
+      sdata->sdata_op_ordering_lock.Unlock();
+      return;
     }
-    break;
-  }
-
-  osd->service.maybe_inject_dispatch_delay();
-
-  // we don't use a Mutex::Locker here because of the
-  // osd->service.release_reserved_pushes() call below
-  sdata->sdata_op_ordering_lock.Lock();
-
-  auto q = sdata->pg_slots.find(token);
-  assert(q != sdata->pg_slots.end());
-  auto& slot = q->second;
-  --slot.num_running;
-
-  if (slot.to_process.empty()) {
-    // raced with wake_pg_waiters or prune_pg_waiters
-    dout(20) << __func__ << " " << token
-	     << " nothing queued" << dendl;
-    if (pg) {
-      pg->unlock();
-    }
-    sdata->sdata_op_ordering_lock.Unlock();
-    return;
-  }
-  if (requeue_seq != slot.requeue_seq) {
-    dout(20) << __func__ << " " << token
-	     << " requeue_seq " << slot.requeue_seq << " > our "
-	     << requeue_seq << ", we raced with wake_pg_waiters"
-	     << dendl;
-    if (pg) {
-      pg->unlock();
+    if (requeue_seq != slot->requeue_seq) {
+      dout(20) << __func__ << " " << token
+	       << " requeue_seq " << slot->requeue_seq << " > our "
+	       << requeue_seq << ", we raced with wake_pg_waiters"
+	       << dendl;
+      if (pg) {
+	pg->unlock();
+      }
+      sdata->sdata_op_ordering_lock.Unlock();
```

Comment:
```_process()``` looks good to me. But next time we update it, let's definitely switch to the standard locking types so we can use ```unique_lock``` and kill *some* of the lock dance around return statements! ;)

Reply:
:+1: